File: client/js/tests/repository.test.ts
import { GovcClient, RepositoryClient } from '../src';
import fetch from 'cross-fetch';

jest.mock('cross-fetch');
const mockFetch = fetch as jest.MockedFunction<typeof fetch>;

describe('RepositoryClient', () => {
  let client: GovcClient;
  let repo: RepositoryClient;

  beforeEach(() => {
    jest.clearAllMocks();
    client = new GovcClient('https://api.example.com');
    repo = new RepositoryClient(client, {
      id: 'test-repo',
      path: '/repos/test-repo',
      current_branch: 'main',
      created_at: new Date().toISOString(),
    });
  });

  describe('file operations', () => {
    it('should add file to staging', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: true,
        status: 200,
        json: async () => ({ status: 'added', message: 'file added' }),
      } as Response);

      await repo.addFile('test.txt', 'Hello, World!');

      expect(mockFetch).toHaveBeenCalledWith(
        'https://api.example.com/api/v1/repos/test-repo/add',
        expect.objectContaining({
          method: 'POST',
          body: JSON.stringify({
            path: 'test.txt',
            content: 'Hello, World!',
          }),
        })
      );
    });

    it('should read file', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: true,
        status: 200,
        json: async () => ({
          path: 'test.txt',
          content: 'Hello, World!',
          size: 13,
        }),
      } as Response);

      const content = await repo.readFile('test.txt');
      
      expect(content).toBe('Hello, World!');
      expect(mockFetch).toHaveBeenCalledWith(
        'https://api.example.com/api/v1/repos/test-repo/read/test.txt',
        expect.any(Object)
      );
    });

    it('should remove file', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: true,
        status: 200,
        json: async () => ({ status: 'removed' }),
      } as Response);

      await repo.removeFile('test.txt');

      expect(mockFetch).toHaveBeenCalledWith(
        'https://api.example.com/api/v1/repos/test-repo/remove/test.txt',
        expect.objectContaining({
          method: 'DELETE',
        })
      );
    });
  });

  describe('git operations', () => {
    it('should create commit', async () => {
      const commitData = {
        hash: 'abc123',
        message: 'Test commit',
        author: 'Test User',
        email: 'test@example.com',
        timestamp: new Date().toISOString(),
      };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        status: 201,
        json: async () => commitData,
      } as Response);

      const commit = await repo.commit('Test commit', {
        name: 'Test User',
        email: 'test@example.com',
      });

      expect(commit).toEqual(commitData);
      expect(mockFetch).toHaveBeenCalledWith(
        'https://api.example.com/api/v1/repos/test-repo/commit',
        expect.objectContaining({
          method: 'POST',
          body: JSON.stringify({
            message: 'Test commit',
            author: 'Test User',
            email: 'test@example.com',
          }),
        })
      );
    });

    it('should get repository status', async () => {
      const statusData = {
        branch: 'main',
        staged: ['file1.txt'],
        modified: ['file2.txt'],
        untracked: ['file3.txt'],
        clean: false,
      };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        status: 200,
        json: async () => statusData,
      } as Response);

      const status = await repo.status();
      
      expect(status).toEqual(statusData);
    });

    it('should get commit log', async () => {
      const commits = [
        {
          hash: 'abc123',
          message: 'Commit 1',
          author: 'User 1',
          email: 'user1@example.com',
          timestamp: new Date().toISOString(),
        },
        {
          hash: 'def456',
          message: 'Commit 2',
          author: 'User 2',
          email: 'user2@example.com',
          timestamp: new Date().toISOString(),
        },
      ];

      mockFetch.mockResolvedValueOnce({
        ok: true,
        status: 200,
        json: async () => ({ commits, count: 2 }),
      } as Response);

      const log = await repo.log(10);
      
      expect(log).toEqual(commits);
      expect(mockFetch).toHaveBeenCalledWith(
        'https://api.example.com/api/v1/repos/test-repo/log?limit=10',
        expect.any(Object)
      );
    });
  });

  describe('branch operations', () => {
    it('should list branches', async () => {
      const branches = [
        { name: 'main', commit: 'abc123', is_current: true },
        { name: 'feature', commit: 'def456', is_current: false },
      ];

      mockFetch.mockResolvedValueOnce({
        ok: true,
        status: 200,
        json: async () => ({ branches, current: 'main', count: 2 }),
      } as Response);

      const result = await repo.listBranches();
      
      expect(result).toEqual(branches);
    });

    it('should create branch', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: true,
        status: 201,
        json: async () => ({ status: 'created', message: 'branch created' }),
      } as Response);

      await repo.createBranch('feature/new', 'main');

      expect(mockFetch).toHaveBeenCalledWith(
        'https://api.example.com/api/v1/repos/test-repo/branches',
        expect.objectContaining({
          method: 'POST',
          body: JSON.stringify({
            name: 'feature/new',
            from: 'main',
          }),
        })
      );
    });

    it('should checkout branch', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: true,
        status: 200,
        json: async () => ({ status: 'success', message: 'switched to feature' }),
      } as Response);

      await repo.checkout('feature');

      expect(repo.currentBranch).toBe('feature');
      expect(mockFetch).toHaveBeenCalledWith(
        'https://api.example.com/api/v1/repos/test-repo/checkout',
        expect.objectContaining({
          method: 'POST',
          body: JSON.stringify({ branch: 'feature' }),
        })
      );
    });
  });

  describe('advanced features', () => {
    it('should create parallel realities', async () => {
      const realities = [
        { name: 'reality-1', isolated: true, ephemeral: true },
        { name: 'reality-2', isolated: true, ephemeral: true },
      ];

      mockFetch.mockResolvedValueOnce({
        ok: true,
        status: 200,
        json: async () => ({ realities, count: 2 }),
      } as Response);

      const result = await repo.createParallelRealities(['reality-1', 'reality-2']);
      
      expect(result).toEqual(realities);
    });

    it('should time travel', async () => {
      const snapshot = {
        time: '2024-01-01T12:00:00Z',
        commit: {
          hash: 'abc123',
          message: 'Historical commit',
          author: 'User',
          email: 'user@example.com',
          timestamp: '2024-01-01T12:00:00Z',
        },
        files: [
          { path: 'file1.txt', content: 'content1' },
          { path: 'file2.txt', content: 'content2' },
        ],
      };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        status: 200,
        json: async () => snapshot,
      } as Response);

      const result = await repo.timeTravel('2024-01-01T12:00:00Z');
      
      expect(result).toEqual(snapshot);
    });
  });

  describe('event streaming', () => {
    it('should create event observable', () => {
      const observable = repo.watch();
      expect(observable).toBeDefined();
    });

    it('should create event stream', () => {
      const stream = repo.createEventStream();
      expect(stream).toBeDefined();
    });
  });
});

File: client/go/govc/repository_test.go
package govc

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"
)

func TestCreateRepo(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			t.Errorf("Expected POST method, got %s", r.Method)
		}
		if r.URL.Path != "/api/v1/repos" {
			t.Errorf("Expected path /api/v1/repos, got %s", r.URL.Path)
		}

		var req struct {
			ID         string `json:"id"`
			MemoryOnly bool   `json:"memory_only"`
		}
		json.NewDecoder(r.Body).Decode(&req)

		if req.ID != "test-repo" {
			t.Errorf("Expected repo ID test-repo, got %s", req.ID)
		}

		repo := Repository{
			ID:            req.ID,
			Path:          "/tmp/test-repo",
			CurrentBranch: "main",
			CreatedAt:     time.Now(),
		}

		w.WriteHeader(http.StatusCreated)
		json.NewEncoder(w).Encode(repo)
	}))
	defer server.Close()

	client, _ := NewClient(server.URL)
	repo, err := client.CreateRepo(context.Background(), "test-repo", &CreateRepoOptions{
		MemoryOnly: true,
	})

	if err != nil {
		t.Fatalf("CreateRepo() error = %v", err)
	}

	if repo.ID != "test-repo" {
		t.Errorf("Expected repo ID test-repo, got %s", repo.ID)
	}
}

func TestRepositoryOperations(t *testing.T) {
	// Create a mock server with multiple endpoints
	mux := http.NewServeMux()

	// Add file endpoint
	mux.HandleFunc("/api/v1/repos/test-repo/add", func(w http.ResponseWriter, r *http.Request) {
		var req struct {
			Path    string `json:"path"`
			Content string `json:"content"`
		}
		json.NewDecoder(r.Body).Decode(&req)

		response := struct {
			Status  string `json:"status"`
			Message string `json:"message"`
		}{
			Status:  "added",
			Message: "file added to staging",
		}
		json.NewEncoder(w).Encode(response)
	})

	// Commit endpoint
	mux.HandleFunc("/api/v1/repos/test-repo/commit", func(w http.ResponseWriter, r *http.Request) {
		var req struct {
			Message string `json:"message"`
			Author  string `json:"author"`
			Email   string `json:"email"`
		}
		json.NewDecoder(r.Body).Decode(&req)

		commit := Commit{
			Hash:      "abc123",
			Message:   req.Message,
			Author:    req.Author,
			Email:     req.Email,
			Timestamp: time.Now(),
		}
		w.WriteHeader(http.StatusCreated)
		json.NewEncoder(w).Encode(commit)
	})

	// Status endpoint
	mux.HandleFunc("/api/v1/repos/test-repo/status", func(w http.ResponseWriter, r *http.Request) {
		status := Status{
			Branch:    "main",
			Staged:    []string{"test.txt"},
			Modified:  []string{},
			Untracked: []string{},
			Clean:     false,
		}
		json.NewEncoder(w).Encode(status)
	})

	// Log endpoint
	mux.HandleFunc("/api/v1/repos/test-repo/log", func(w http.ResponseWriter, r *http.Request) {
		commits := []*Commit{
			{
				Hash:      "abc123",
				Message:   "Initial commit",
				Author:    "Test User",
				Email:     "test@example.com",
				Timestamp: time.Now(),
			},
		}
		response := struct {
			Commits []*Commit `json:"commits"`
			Count   int       `json:"count"`
		}{
			Commits: commits,
			Count:   len(commits),
		}
		json.NewEncoder(w).Encode(response)
	})

	server := httptest.NewServer(mux)
	defer server.Close()

	client, _ := NewClient(server.URL)
	repo := &Repository{
		client: client,
		ID:     "test-repo",
	}

	t.Run("AddFile", func(t *testing.T) {
		err := repo.AddFile(context.Background(), "test.txt", "Hello, World!")
		if err != nil {
			t.Errorf("AddFile() error = %v", err)
		}
	})

	t.Run("Commit", func(t *testing.T) {
		commit, err := repo.Commit(context.Background(), "Initial commit", &Author{
			Name:  "Test User",
			Email: "test@example.com",
		})
		if err != nil {
			t.Fatalf("Commit() error = %v", err)
		}
		if commit.Hash != "abc123" {
			t.Errorf("Expected commit hash abc123, got %s", commit.Hash)
		}
	})

	t.Run("Status", func(t *testing.T) {
		status, err := repo.Status(context.Background())
		if err != nil {
			t.Fatalf("Status() error = %v", err)
		}
		if status.Branch != "main" {
			t.Errorf("Expected branch main, got %s", status.Branch)
		}
		if len(status.Staged) != 1 {
			t.Errorf("Expected 1 staged file, got %d", len(status.Staged))
		}
	})

	t.Run("Log", func(t *testing.T) {
		commits, err := repo.Log(context.Background(), 10)
		if err != nil {
			t.Fatalf("Log() error = %v", err)
		}
		if len(commits) != 1 {
			t.Errorf("Expected 1 commit, got %d", len(commits))
		}
	})
}

func TestBranchOperations(t *testing.T) {
	mux := http.NewServeMux()

	// List branches
	mux.HandleFunc("/api/v1/repos/test-repo/branches", func(w http.ResponseWriter, r *http.Request) {
		if r.Method == http.MethodGet {
			branches := []*Branch{
				{Name: "main", Commit: "abc123", IsCurrent: true},
				{Name: "feature", Commit: "def456", IsCurrent: false},
			}
			response := struct {
				Branches []*Branch `json:"branches"`
				Current  string    `json:"current"`
				Count    int       `json:"count"`
			}{
				Branches: branches,
				Current:  "main",
				Count:    len(branches),
			}
			json.NewEncoder(w).Encode(response)
		} else if r.Method == http.MethodPost {
			// Create branch
			var req struct {
				Name string `json:"name"`
				From string `json:"from"`
			}
			json.NewDecoder(r.Body).Decode(&req)

			w.WriteHeader(http.StatusCreated)
			json.NewEncoder(w).Encode(struct {
				Status  string `json:"status"`
				Message string `json:"message"`
			}{
				Status:  "created",
				Message: "branch created",
			})
		}
	})

	// Checkout
	mux.HandleFunc("/api/v1/repos/test-repo/checkout", func(w http.ResponseWriter, r *http.Request) {
		var req struct {
			Branch string `json:"branch"`
		}
		json.NewDecoder(r.Body).Decode(&req)

		json.NewEncoder(w).Encode(struct {
			Status  string `json:"status"`
			Message string `json:"message"`
		}{
			Status:  "success",
			Message: "switched to branch " + req.Branch,
		})
	})

	server := httptest.NewServer(mux)
	defer server.Close()

	client, _ := NewClient(server.URL)
	repo := &Repository{
		client: client,
		ID:     "test-repo",
	}

	t.Run("ListBranches", func(t *testing.T) {
		branches, err := repo.ListBranches(context.Background())
		if err != nil {
			t.Fatalf("ListBranches() error = %v", err)
		}
		if len(branches) != 2 {
			t.Errorf("Expected 2 branches, got %d", len(branches))
		}
	})

	t.Run("CreateBranch", func(t *testing.T) {
		err := repo.CreateBranch(context.Background(), "new-feature", "main")
		if err != nil {
			t.Errorf("CreateBranch() error = %v", err)
		}
	})

	t.Run("Checkout", func(t *testing.T) {
		err := repo.Checkout(context.Background(), "feature")
		if err != nil {
			t.Errorf("Checkout() error = %v", err)
		}
		if repo.CurrentBranch != "feature" {
			t.Errorf("Expected current branch to be feature, got %s", repo.CurrentBranch)
		}
	})
}
